tiledb/sm/file/file.cc

```diff
@@ -0,0 +1,664 @@
+  //  Array::open(query_type, encryption_type, encryption_key, key_length);
+  return Array::open(
+      query_type,
+      timestamp_start_,
+      timestamp_end_,
+      encryption_type,
+      encryption_key,
+      key_length);
+}
+
+Status File::open(
+    QueryType query_type,
+    uint64_t timestamp_start,
+    uint64_t timestamp_end,
+    EncryptionType encryption_type,
+    const void* encryption_key,
+    uint32_t key_length) {
+  return Array::open(
+      query_type,
+      timestamp_start,
+      timestamp_end,
+      encryption_type,
+      encryption_key,
+      key_length);
+}
+
+Status File::create([[maybe_unused]] const Config* config) {
+  try {
+    auto encryption_key = get_encryption_key_from_config(config_);
+    RETURN_NOT_OK(storage_manager_->array_create(
+        array_uri_, &file_schema_, *encryption_key));
+  } catch (const std::exception& e) {
+    return Status::FileError(e.what());
+  }
+
+  return Status::Ok();
+}
+
+Status File::create_from_uri(const URI& file, const Config* config) {
+  try {
+    VFS vfs;
+    // Initialize VFS object
+    auto stats = storage_manager_->stats();
+    auto compute_tp = storage_manager_->compute_tp();
+    auto io_tp = storage_manager_->io_tp();
+    auto vfs_config = config ? config : nullptr;
+    auto ctx_config = storage_manager_->config();
+    RETURN_NOT_OK(vfs.init(stats, compute_tp, io_tp, &ctx_config, vfs_config));
+
+    VFSFileHandle vfsfh(file, &vfs, VFSMode::VFS_READ);
+
+    RETURN_NOT_OK(create_from_vfs_fh(&vfsfh, config));
+    RETURN_NOT_OK(vfsfh.close());
+    RETURN_NOT_OK(vfs.terminate());
+  } catch (const std::exception& e) {
+    return Status::FileError(e.what());
+  }
+  return Status::Ok();
+}
+
+Status File::create_from_vfs_fh(
+    const VFSFileHandle* file, [[maybe_unused]] const Config* config) {
+  try {
+    if (file->mode() != VFSMode::VFS_READ)
+      return Status::FileError("File must be open in READ mode");
+
+    uint64_t size = file->size();
+    file_schema_.set_schema_based_on_file_details(size, false);
+    auto encrpytion_key = get_encryption_key_from_config(config_);
+    RETURN_NOT_OK(storage_manager_->array_create(
+        array_uri_, &file_schema_, *encrpytion_key));
+  } catch (const std::exception& e) {
+    return Status::FileError(e.what());
+  }
+
+  return Status::Ok();
+}
+
+Status File::save_from_file_handle(FILE* in, const Config* config) {
+  try {
+    if (query_type_ != QueryType::WRITE)
+      return Status::FileError(
+          "Can not save file; File opened in read mode not write mode");
+
+    // Get file size
+    fseek(in, 0L, SEEK_END);
+    uint64_t size = ftell(in);
+    rewind(in);
+
+    // TODO: Add config option to let the user control how much of the file we
+    // we read
+    // We can support partial writes either global order (single fragment)
+    // or row-major with multiple fragment but same timestamp
+    Buffer buffer;
+    buffer.realloc(size);
+    size = fread(buffer.data(), 1, size, in);
```

Comment:
TODOs here to check
- whether the file is readable (could be WRONLY)
- `ferror` and `feof` to make sure we actually read the whole thing